File: test/src/FileParser.java
import java.util.Scanner;
import java.io.*;
import java.util.*;

public class FileParser {
    private String inputfile;
    private String outputfile;
    private int digitosPalabra;
    private ArrayList<String> events;
    private TreeSet<String> palabras;
    private HashMap<String, Integer> frec;
    private HashMap<String, Double> prob;
    private HashMap<String, Double> cantInfo;

    public FileParser(String inputfile, int digitosPalabra, String outputfile) {
        this.inputfile = inputfile;
        this.digitosPalabra = digitosPalabra;
        this.outputfile = outputfile;
    }

    public void parseFile() throws FileNotFoundException, IOException {
        File file = null;
        FileReader fileReader = null;
        BufferedReader bufferedReader = null;
        Scanner scan = null;
        String regex;

        file = new File(this.inputfile);
        fileReader = new FileReader(file);
        bufferedReader = new BufferedReader(fileReader);
        scan = new Scanner(bufferedReader);

        regex = "(?<=\\G.{" + String.valueOf(this.digitosPalabra) + "})";
        String[] arr = scan.nextLine().split(regex);

        scan.close();
        bufferedReader.close();
        fileReader.close();

        this.events = new ArrayList<String>(Arrays.asList(arr));
        this.frec = new HashMap<>();
        for (String str : this.events) {
            this.frec.put(str, this.frec.getOrDefault(str, 0) + 1);
        }
        this.palabras = new TreeSet<String>(this.frec.keySet());
        this.prob = new HashMap<>();
        for (String str : this.palabras) {
            this.prob.put(str, (double) this.frec.get(str) / this.events.size());
        }
        this.cantInfo = new HashMap<>();
        for (String str : this.palabras) {
            this.cantInfo.put(str, cantidadDeInformacion(this.prob.get(str)));
        }
    }

    public void printFrecuencies() {
        System.out.println("Frecuencias: ");

        for (String s : this.palabras) {
            System.out.println("[" + s + "] = " + this.frec.get(s));
        }
        System.out.println();
    }

    public void printCantidadDeInfo() {
        System.out.println("Cantidad de info: ");

        for (String s : this.palabras) {
            System.out.println("I(" + s + ") = " + this.cantInfo.get(s));
        }
        System.out.println();
    }

    public void printProbabilidad() {
        double total = 0;
        System.out.println("Probabilidades: ");

        for (String s : this.palabras) {
            System.out.println("P(" + s + ") = " + this.prob.get(s));
            total += this.prob.get(s);
        }
        System.out.println("sum = " + total);
        System.out.println();
        // System.out.println("sum = " +
        // this.prob.values().stream().mapToDouble(Double::doubleValue).sum());
    }

    private double cantidadDeInformacion(double prob) {
        return -1 * Math.log(prob) / Math.log(2);
    }

    public double entropia() {
        double h = 0;
        for (String s : this.palabras) {
            h += this.prob.get(s) * this.cantInfo.get(s);
        }
        return h;
    }

    public void writeToCsv() throws IOException {
        Writer file = new FileWriter(this.outputfile);

        file.write(String.join(",", "Palabras", "Frecuencia", "Probabilidad", "Cant. Informacion", "\n"));
        for (String key : this.palabras) {
            file.write(String.join(",", key, Integer.toString(this.frec.get(key)), Double.toString(this.prob.get(key)),
                    Double.toString(this.cantInfo.get(key)), "\n"));
        }
        file.write(String.join(",", "Entropia", Double.toString(this.entropia())));
        file.close();
    }

    public static void main(String[] args) {
        try {
            FileParser fp = new FileParser(args[0], Integer.valueOf(args[1]), args[2]);
            fp.parseFile();
            fp.printFrecuencies();
            fp.printProbabilidad();
            fp.printCantidadDeInfo();
            System.out.println("H(S) = " + fp.entropia());
            fp.writeToCsv();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException ioe) {
            ioe.printStackTrace();
        }
    }
}

File: nousado/src/fileparser.py
import sys
import re
import math
from pprint import pprint


class FileParser:
    def __init__(self, inputfile, digitos_palabra):
        self.__inputfile = inputfile
        self.__outputfile = None
        self.__digitos_palabra = digitos_palabra
        self.__events = {}
        self.__frec = {}
        self.__prob = {}
        self.__cant_info = {}


    def parsefile(self):
        regex = "." * self.__digitos_palabra
        with open(self.__inputfile, 'r') as txt:
            file_content = txt.read()
        self.__events = re.findall(regex, file_content)
        for e in self.__events:
            self.__frec[e] = self.__frec.get(e, 0) + 1
        total = sum(self.__frec.values())
        for key in self.__frec.keys():
            self.__prob[key] = self.__frec[key] / total
        for key in self.__frec.keys():
            self.__cant_info[key] = FileParser.cant_info(self.__prob[key])

    
    def entropia(self) -> float:
        h = 0
        for event in self.__cant_info.keys():
            h += self.__prob[event] * self.__cant_info[event]
        return h


    def getCantInfo(self):
        return self.__cant_info


    def cant_info(prob: float) -> float:
        return math.log(prob, 2) * -1



def main(args):
    fileparser = FileParser(args[1], int(args[2]))
    fileparser.parsefile()
    pprint(fileparser.getCantInfo())
    print(f"H(s) = {fileparser.entropia()}")


if __name__ == '__main__':
    main(sys.argv)

